### PR TITLE
initial version of spec for key backups

### PR DESCRIPTION
**Pull request by uhoreg**
Spec PR for #1219 / #1538 and #2526.

Does not have any mentions of SSSS or cross-signing keys, since those parts are not written up yet.

Rendered version of changes at https://12751-24998719-gh.circle-artifacts.com/0/scripts/gen/client_server/unstable.html#server-side-key-backups

### CODE REVIEW

File: api/client-server/definitions/key_backup_data.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+description: "The key data"
+properties:
+  first_message_index:
+    description: |-
+      The index of the first message in the session that the key can decrypt.
+    type: integer
+    example: 1
+  forwarded_count:
+    description: |-
+      The number of times this key has been forwarded.
```
Comment (ara4n) [inline api/client-server/definitions/key_backup_data.yaml]:
```suggestion
      The number of times this key has been forwarded via key-sharing between devices.
```

File: api/client-server/definitions/key_backup_data.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    type: integer
+    example: 0
+  is_verified:
+    description: |-
+      Whether the device backing up the key was verified the device that the key
```
Comment (ara4n) [inline api/client-server/definitions/key_backup_data.yaml]:
```suggestion
      Whether the device backing up the key verified the device that the key
```

File: api/client-server/key_backup.yaml [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+                example: {
+                    "public_key": "abcdefg",
+                    "signatures": {
+                      "@alice:example.org": {
+                        "ed25519:deviceid": "signature"
+                      }
+                    }
+                  }
+              count:
+                description: The number of key stored in the backup.
```
Comment (ara4n) [inline api/client-server/key_backup.yaml]:
```suggestion
                description: The number of keys stored in the backup.
```

File: api/client-server/key_backup.yaml [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+                      }
+                    }
+                  }
+            required:
+              - algorithm
+              - auth_data
+      responses:
+        200:
+          description:
+            The version of the new backup.
```
Comment (ara4n) [inline api/client-server/key_backup.yaml]:
```suggestion
            The version id of the new backup.
```

File: specification/modules/end_to_end_encryption.rst [hunk 5]
```diff
@@ -767,6 +767,126 @@ previously-received ``request`` message with the same ``request_id`` and
...
+stored on the server encrypted, it requires users to enter a decryption key to
+decrypt the session keys.
+
+To create a backup, a client will call ``POST /room_keys/version`` and define
+how the keys are to be encrypted through the backup's ``auth_data``; other
+clients can discover the backup by calling ``GET /room_keys/version``.  Keys
+are encrypted according to the backups ``auth_data`` and added to the backup by
+calling ``PUT /room_keys/keys?version=$v`` or one of its variants, and can be
+retrieved by calling ``GET /room_keys/keys?version=$v`` or one of its variants.
+Backups can also be deleted using ``DELETE /room_keys/version``, or individual
```
Comment (ara4n) [inline specification/modules/end_to_end_encryption.rst]:
```suggestion
Keys can only be written to the most recently created version of the backup.
Backups can also be deleted using ``DELETE /room_keys/version``, or individual
```

File: specification/modules/end_to_end_encryption.rst [hunk 6]
```diff
@@ -767,6 +767,126 @@ previously-received ``request`` message with the same ``request_id`` and
...
+   as the info.  The first 32 bytes are used as the AES key, the next 32 bytes
+   are used as the MAC key, and the last 16 bytes are used as the AES
+   initialization vector.
+4. Stringify the JSON object, and encrypt it using AES-CBC-256 with PKCS#7
+   padding.  This encrypted data, encoded using unpadded base64, becomes the
+   ``ciphertext`` property of the ``session_data``.
+5. Pass the raw encrypted data (prior to base64 encoding) through HMAC-SHA-256
+   using the MAC key generated above.  The first 8 bytes of the resulting MAC
+   are base64-encoded, and become the ``mac`` property of the ``session_data``.
+
```
Comment (ara4n) [inline specification/modules/end_to_end_encryption.rst]: I assume this is similar to the key export format? If so, we should probably acknowledge that to aid comprehension.

**Comment by ara4n** [comment]
lgtm from medium-to-high level review :)

File: specification/modules/end_to_end_encryption.rst (hunk 6 above)
Comment (uhoreg) [reply specification/modules/end_to_end_encryption.rst]: Added a note

File: api/client-server/definitions/key_backup_data.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline api/client-server/definitions/key_backup_data.yaml]: 😢 that this has been stuck in the queue this long

File: api/client-server/key_backup.yaml [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+produces:
+  - application/json
+securityDefinitions:
+  $ref: definitions/security.yaml
+paths:
+  "/room_keys/version":
+    post:
+      summary: Create a new backup.
+      description: |-
+        Creates a new backup
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
        Creates a new backup.
```

File: api/client-server/key_backup.yaml [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+      operationId: postRoomKeysVersion
+      security:
+        - accessToken: []
+      parameters:
+        - in: body
+          name: version
+          description: "The backup configuration"
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
          description: "The backup configuration."
```

File: api/client-server/key_backup.yaml [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+                      "@alice:example.org": {
+                        "ed25519:deviceid": "signature"
+                      }
+                    }
+                  }
+            required:
+              - algorithm
+              - auth_data
+      responses:
+        200:
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]: Applies to all endpoints: do we have rate limiting (429) on this? are there common errors people need to watch out for (besides invalid json, etc)?

File: api/client-server/key_backup.yaml [hunk 11]
```diff
@@ -0,0 +1,761 @@
...
+          description:
+            The version id of the new backup.
+          schema:
+            type: object
+            properties:
+              version:
+                type: string
+                description: The backup version
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
                description: The backup version.
```

File: api/client-server/key_backup.yaml [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+      operationId: getRoomKeysVersion
+      security:
+        - accessToken: []
+      parameters:
+        - in: path
+          type: string
+          name: version
+          description: |-
+            The backup version to get. When an empty string, the trailing slash
+            is optional, and the current backup is returned.
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]: Is there a format to these?

File: api/client-server/key_backup.yaml [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+                example: "1"
+            required:
+              - algorithm
+              - auth_data
+              - count
+              - etag
+              - version
+        404:
+          description:
+            The backup specified does not exist
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
            The backup specified does not exist.
```

File: api/client-server/key_backup.yaml [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+          examples:
+            application/json: {
+                "errcode": "M_NOT_FOUND",
+                "error": "Unknown backup version"
+              }
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]: The indentation this was probably copied from is an autoconvert fail from years ago
```suggestion
              "errcode": "M_NOT_FOUND",
              "error": "Unknown backup version"
            }
```

File: api/client-server/key_backup.yaml [hunk 15]
```diff
@@ -0,0 +1,761 @@
...
+        Update information about an existing backup.  Only ``auth_data`` can be modified.
+      operationId: putRoomKeysVersion
+      security:
+        - accessToken: []
+      parameters:
+        - in: path
+          type: string
+          name: version
+          description: |-
+            The backup version to update.
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]: We should repeat requirements for the version schema here too.

File: api/client-server/key_backup.yaml [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+                  expected format of the data.
+                type: object
+                example: {
+                    "public_key": "abcdefg",
+                    "signatures": {
+                      "@alice:example.org": {
+                        "ed25519:deviceid": "signature"
+                      }
+                    }
+                  }
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]: looks like indentation in general is a bit wonky throughout the PR

File: api/client-server/key_backup.yaml [hunk 17]
```diff
@@ -0,0 +1,761 @@
...
+                  The backup version.  If present, must be the same as the
+                  version in the path parameter.
+                type: string
+                example: "1"
+            required:
+              - algorithm
+              - auth_data
+      responses:
+        200:
+          description: The update succeeded
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
          description: The update succeeded.
```

File: api/client-server/key_backup.yaml [hunk 18]
```diff
@@ -0,0 +1,761 @@
...
+          schema:
+            type: object
+            properties: {}
+        404:
+          description: The backup specified does not exist
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
          description: The backup specified does not exist.
```

File: api/client-server/key_backup.yaml [hunk 19]
```diff
@@ -0,0 +1,761 @@
...
+                "errcode": "M_NOT_FOUND",
+                "error": "Unknown backup version"
+              }
+          schema:
+            "$ref": "definitions/errors/error.yaml"
+      tags:
+        - End-to-end encryption
+  "/room_keys/keys/{roomId}/{sessionId}":
+    put:
+      summary: Store a key in the backup
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
      summary: Store a key in the backup.
```

File: api/client-server/key_backup.yaml [hunk 20]
```diff
@@ -0,0 +1,761 @@
...
+          x-example: "!roomid:example.org"
+        - in: path
+          type: string
+          name: sessionId
+          description: The ID of the megolm session that the key is for.
+          required: true
+          x-example: "sessionid"
+        - in: body
+          name: data
+          description: "The key data"
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
          description: "The key data."
```

File: api/client-server/key_backup.yaml [hunk 21]
```diff
@@ -0,0 +1,761 @@
...
+          schema:
+            "$ref": "definitions/key_backup_data.yaml"
+      responses:
+        200:
+          description: The update succeeded
```
Comment (turt2live) [inline api/client-server/key_backup.yaml]:
```suggestion
          description: The update succeeded.
```


... and I think you get the idea :p

File: specification/modules/end_to_end_encryption.rst [hunk 22]
```diff
@@ -767,6 +767,129 @@ previously-received ``request`` message with the same ``request_id`` and
...
+presented to the user to save, it is presented as a string constructed as
+follows:
+
+1. The 256-bit curve25519 private key is prepended by the bytes ``0x8B`` and
+   ``0x01``
+2. All the bytes in the string above, including the two header bytes, are XORed
+   together to form a parity byte. This parity byte is appended to the byte
+   string.
+3. The byte string is encoded using base58, using the same mapping as is used
+   for Bitcoin addresses.
```
Comment (turt2live) [inline specification/modules/end_to_end_encryption.rst]: for clarity, can we link to/include what this is?

File: api/client-server/key_backup.yaml (hunk 12 above)
Comment (uhoreg) [reply api/client-server/key_backup.yaml]: I think I've addressed this in the latest changes.

File: api/client-server/key_backup.yaml (hunk 16 above)
Comment (uhoreg) [reply api/client-server/key_backup.yaml]: Hopefully I caught them all.

**Comment by uhoreg** [comment]
Waiting for https://github.com/matrix-org/matrix-doc/pull/2526, and I still need to check error codes.

File: api/client-server/key_backup.yaml (hunk 10 above)
Comment (uhoreg) [reply api/client-server/key_backup.yaml:68]: Added an error code.  Synapse doesn't have rate limiting on anything, currently.  I don't see a particular need for rate limiting on these endpoints, other than the general "everything should be rate limited".

File: api/client-server/key_backup.yaml [hunk 23]
```diff
@@ -0,0 +1,816 @@
...
+        - accessToken: []
+      parameters:
+        - in: path
+          type: string
+          name: version
+          description: |-
+            The backup version to get, as returned in the ``version`` parameter
+            of the response in `POST /_matrix/client/r0/room_keys/version`_ or
+            this endpoint. When an empty string, the trailing slash is
+            optional, and the current backup is returned.
```
Comment (richvdh) [inline api/client-server/key_backup.yaml:98]: This touches on #2107, but I'm not a fan of treating "empty string" the same as "missing param". I feel like we went through all this fairly recently with some other endpoints, though I can't currently quite remember what. something to do with `/_matrix/client/r0/rooms/{roomId}/state` maybe?

Comment (turt2live) [reply api/client-server/key_backup.yaml:98]: For `/state` we decided that `/state/` and `/state` were the same, as the state key is implied to be `""`.

Comment (richvdh) [reply api/client-server/key_backup.yaml:98]: itym `/state/{type}` and `/state/{type}/` ?

well that seems to go against my argument, but still, I'd prefer us not to propagate the error... there is prior art in the other direction, fwiw.

Comment (turt2live) [reply api/client-server/key_backup.yaml:98]: yes, sorry.

There's prior art in all the directions we have taken a look at:
* Optional slash (state keys, etc)
* Mandatory slash (push rules)
* Mandatory not-slash (a large chunk of the spec)

Comment (uhoreg) [reply api/client-server/key_backup.yaml]: Looking at this again, we actually never supported having a trailing slash and empty version.  So I've removed reference to this.

File: specification/modules/end_to_end_encryption.rst [hunk 24]
```diff
@@ -780,15 +780,18 @@ device to be online from which to request keys. However, as the session keys are
...
-retrieved by calling ``GET /room_keys/keys?version=$v`` or one of its variants.
-Keys can only be written to the most recently created version of the backup.
-Backups can also be deleted using ``DELETE /room_keys/version``, or individual
-keys can be deleted using ``DELETE /room_key/keys?version=$v`` or one of its
+To create a backup, a client will call `POST
+/_matrix/client/r0/room_keys/version`_ and define how the keys are to be
+encrypted through the backup's ``auth_data``; other clients can discover the
+backup by calling `GET /_matrix/client/r0/room_keys/version/{version}`_,
+setting ``{version}`` to the empty string.  Keys are encrypted according to the
+backups ``auth_data`` and added to the backup by calling `PUT
```
Comment (turt2live) [inline specification/modules/end_to_end_encryption.rst:788]:
```suggestion
backup's ``auth_data`` and added to the backup by calling `PUT
```

File: api/client-server/key_backup.yaml [hunk 25]
```diff
@@ -0,0 +1,884 @@
+# Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline api/client-server/key_backup.yaml:1]: Presumably also 2020 at this point:
```suggestion
# Copyright 2019-2020 The Matrix.org Foundation C.I.C.
```

File: specification/modules/end_to_end_encryption.rst [hunk 26]
```diff
@@ -767,6 +767,134 @@ previously-received ``request`` message with the same ``request_id`` and
...
+In contrast with `Key requests`_, Server-side key backups do not require another
+device to be online from which to request keys. However, as the session keys are
+stored on the server encrypted, it requires users to enter a decryption key to
+decrypt the session keys.
+
+To create a backup, a client will call `POST
+/_matrix/client/r0/room_keys/version`_ and define how the keys are to be
+encrypted through the backup's ``auth_data``; other clients can discover the
+backup by calling `GET /_matrix/client/r0/room_keys/version/{version}`_,
+setting ``{version}`` to the empty string.  Keys are encrypted according to the
```
Comment (turt2live) [inline specification/modules/end_to_end_encryption.rst:787]: This is now using the `/room_keys/version` (no empty param/slash) endpoint?

File: api/client-server/key_backup.yaml (hunk 10 above)
Comment (turt2live) [reply api/client-server/key_backup.yaml:68]: At the moment the spec renders as a bunch of "Rate limited: No" endpoints, which is probably not intended. To make them rate limited, copy/paste the 429 error from another endpoint and it'll happily update.

**Pull request merged**